uaclient/api/u/pro/token_info/v1.py

```diff
@@ -0,0 +1,246 @@
+from uaclient.api.api import APIEndpoint
+from uaclient.api.data_types import AdditionalInfo
+from uaclient.config import UAConfig
+from uaclient.contract import get_available_resources, get_contract_information
+from uaclient.data_types import (
+    BoolDataValue,
+    DataObject,
+    Field,
+    StringDataValue,
+    data_list,
+)
+from uaclient.entitlements import entitlement_factory
+
+LOG = logging.getLogger(util.replace_top_level_logger_name(__name__))
+
+
+class TokenInfoOptions(DataObject):
+    fields = [
+        Field(
+            "token", StringDataValue, doc="Token to be used for the request"
+        ),
+    ]
+
+    def __init__(self, token: str):
+        self.token = token
+
+
+class AccountInfo(DataObject):
+    fields = [
+        Field("id", StringDataValue),
+        Field("name", StringDataValue),
+    ]
+
+    def __init__(self, id: str, name: str):
+        self.id = id
+        self.name = name
+
+
+class ContractInfo(DataObject):
+    fields = [
+        Field("id", StringDataValue),
+        Field("name", StringDataValue),
+    ]
+
+    def __init__(self, id: str, name: str):
+        self.id = id
+        self.name = name
+
+
+class ServiceInfo(DataObject):
+    fields = [
+        Field("name", StringDataValue),
+        Field("description", StringDataValue),
+        Field("entitled", BoolDataValue),
+        Field("auto_enabled", BoolDataValue),
+        Field("available", BoolDataValue),
+    ]
+
+    def __init__(
+        self,
+        available: bool,
+        description: str,
+        entitled: bool,
+        name: str,
+        auto_enabled: bool,
+    ):
+        self.available = available
+        self.description = description
+        self.entitled = entitled
+        self.name = name
+        self.auto_enabled = auto_enabled
+
+
+class TokenInfoResult(DataObject, AdditionalInfo):
+    fields = [
+        Field("account", AccountInfo),
+        Field("contract", ContractInfo),
+        Field("services", data_list(ServiceInfo)),
+    ]
+
+    def __init__(
+        self,
+        account: AccountInfo,
+        contract: ContractInfo,
+        services: List[ServiceInfo],
+    ):
+        self.account = account
+        self.contract = contract
+        self.services = services
+
+
+def _get_entitlement_information_bool(
+    entitlements: List[Dict[str, Any]], entitlement_name: str
+) -> Dict[str, Any]:
+    """Extract information from the entitlements array."""
+    for entitlement in entitlements:
+        if entitlement.get("type") == entitlement_name:
+            return {
+                "entitled": entitlement.get("entitled"),
+                "auto_enabled": (
+                    entitlement.get("obligations", {}).get("enableByDefault")
+                ),
+            }
+    return {"entitled": False, "auto_enabled": False}
+
+
+def get_token_info(options: TokenInfoOptions) -> TokenInfoResult:
+    return _get_token_info(options=options, cfg=UAConfig())
+
+
+def _get_token_info(
```

Comment:
Can we reuse this API for `status --simulate-with-token` ?
That is because the logic here is really similar to what we have on status and that would avoid code duplication

Reply:
I was looking to do that initially but because of how the difference between the structure of we currently return `simulated_status` and how the api returns it, we might break the way we generate status.json.

Integrating the api is also diffifcult since we will be using the same function calls either ways to keep the current structure and everything will just end up being called twice. 
Is there a better way to handle this?

Reply:
I would say this could be reused there, yes.

One thing that I am missing here is the expiration check - in the simulated status, we error out when the contract is expired or not yet valid `status.py:465-493`. The spec states that this situation should cause the API to return an error too.

Once that is in place, calling the API in status will:
- delegate the token checks to the API code
- return the services list in a state we can use to build the outputs

I understand it is not straightforward as other endpoints would be, but I do think we should do our best to reuse things from here.

Reply:
The part I am a little skeptical about is the gain is not much here given the token_info has less information than needed for the `simulate_status`. Even if we delegate the token checks, we would still require the `get_contract_information()` output either ways since `token_info` does not have all the contract and account information we need. 

A better way to handle this is maybe ->

- Use a different function in the api file to get the contract and account information, use that function in both the api and the `simulate_status` function OR add a flag to api function to return `all` information which defaults to False (just the `_get_token_info` function)
- Move the `expiration` and contract check to the API and use the API in `simulate_status`

This could completely replace the `simulate_status` function code and we can just the api there. What do you guys think about this? @lucasmoura @renanrodrigo

Reply:
Adding to the above comment -
Currently the fields that are included in the `simualte_status` response which are not present in the api are:
```
contract: {
  "created_at",
  "products"
}
account: {
  "created_at",
  "external_account_ids"
}
"expires"
"effectiveTo"
```

We do not want to query the `subscription` api with the token, as @renanrodrigo said that should just work when asking for subscription information.

Reply:
@dheyay I agree that some of those fields are not necessary for this particular API, but maybe including the `expires` and `effectiveTo` are actually a good idea, since I believe users might be interested in that information.

But I also agree that then, this new API is not enough to be used by `--simulate-with-token`

Reply:
@lucasmoura @renanrodrigo @orndorffgrant Should we add `expires` and `effectiveTo` to the token_info api

Reply:
Updated with the new fields @lucasmoura